[Symmetric memory] Polish NCCL symm mem code

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170582

Some cleanups are needed for nccl backend, while debugging double memory release issue.

fduwjj committed:
[Symmetric memory] Polish NCCL symm mem code

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -386,9 +383,6 @@ class NCCLSymmetricMemoryAllocator : public SymmetricMemoryAllocator {
-  std::unordered_map<void*, c10::intrusive_ptr<SymmetricMemory>>
-      ptr_to_symm_mem_;
-
   std::unordered_map<void*, std::shared_ptr<NCCLAllocation>> allocations_;
```
Can this be a unique ptr too now?

fduwjj replied:
Opps, yes.

fduwjj committed:
Update on "[Symmetric memory] Polish NCCL symm mem code"




[ghstack-poisoned]

fduwjj committed:
Update on "[Symmetric memory] Polish NCCL symm mem code"


Some cleanups are needed for nccl backend, while debugging double memory release issue.


[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -67,20 +67,19 @@ class NCCLSymmetricMemory : public SymmetricMemory {
  NCCLSymmetricMemory(
-      std::shared_ptr<NCCLAllocation> allocation,
+      NCCLAllocation* allocation,
       const std::string& group_name,
```
pass by value is also a copy right? then my question is what is benefit of doing this just for my learning purpose?

fduwjj committed:
Update on "[Symmetric memory] Polish NCCL symm mem code"


Some cleanups are needed for nccl backend, while debugging double memory release issue.


[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge -i

Skylion007 commented:
@fduwjj Caught a potential bug we should patch in the the allocation struct: https://github.com/pytorch/pytorch/issues/170894
